### PR TITLE
Fix command-line arguments & allow using env vars and wildcard IP addresses for server

### PR DESCRIPTION
**Pull request by Pierstoval**
Using PHP 7.2.2 on Windows, testing this code in the WebServerManager:

```php
$commandLine = [$binary] + $finder->findArguments() + ['-dvariables_order=EGPCS', '-S', \sprintf('%s:%d', $this->hostname, $this->port)];

$commandLine2 = array_merge([$binary], $finder->findArguments(), ['-dvariables_order=EGPCS', '-S', \sprintf('%s:%d', $this->hostname, $this->port)]);

dump($commandLine);
dump($commandLine2);
```

And I have this output:

```
array:3 [
  0 => "E:\dev\php72-nts\php.exe"
  1 => "-S"
  2 => "0.0.0.0:12000"
]
array:4 [
  0 => "E:\dev\php72-nts\php.exe"
  1 => "-dvariables_order=EGPCS"
  2 => "-S"
  3 => "0.0.0.0:12000"
]
```

The first one is wrong, so I decided to refactor it with a plain `array_merge()` that will surely be working 👍 

This PR also introduces lots of other changes that allow using env vars and ips like `0.0.0.0` for servers listening to any host. I thought it would be a really nice improvement

### CODE REVIEW

**Comment by dunglas** [comment]
Can you provide the bug fix in a separate PR please?

File: src/ProcessManager/WebServerReadinessProbeTrait.php [hunk 1]
```diff
@@ -36,23 +36,46 @@ private function checkPortAvailable(string $hostname, int $port, bool $throw = t
         }
     }
 
-    public function waitUntilReady(Process $process, string $url, bool $ignoreErrors = false): void
+    public function waitUntilReady(Process $process, string $url): void
     {
-        $context = \stream_context_create(['http' => [
```
Comment (dunglas) [inline src/ProcessManager/WebServerReadinessProbeTrait.php]: Why this change? It will not work if the `/` route isn't defined for instance.

File: src/ProcessManager/WebServerReadinessProbeTrait.php [hunk 2]
```diff
@@ -36,23 +36,46 @@ private function checkPortAvailable(string $hostname, int $port, bool $throw = t
...
-            'ignore_errors' => $ignoreErrors,
-            'protocol_version' => '1.1',
-            'header' => ['Connection: close'],
-            'timeout' => 1,
-        ]]);
-
-        while (Process::STATUS_STARTED !== ($status = $process->getStatus()) || false === @\file_get_contents($url, false, $context)) {
+        $host = parse_url($url, PHP_URL_HOST);
+
+        if ($host === '0.0.0.0') {
```
Comment (dunglas) [inline src/ProcessManager/WebServerReadinessProbeTrait.php]: Please use yoda style.

File: src/ProcessManager/WebServerReadinessProbeTrait.php [hunk 3]
```diff
@@ -36,23 +36,46 @@ private function checkPortAvailable(string $hostname, int $port, bool $throw = t
...
 
+            if ($errno !== 0) {
+                $socketErrors[] = "#$errno:$errstr";
+            }
+
             // block until the web server is ready
             \usleep(1000);
+        } while (Process::STATUS_STARTED !== $status || ++$retries === $maxRetries);
+
+        if (count($socketErrors)) {
```
Comment (dunglas) [inline src/ProcessManager/WebServerReadinessProbeTrait.php]: `\count()`

File: src/ProcessManager/WebServerReadinessProbeTrait.php (hunk 1 above)
Comment (Pierstoval) [reply src/ProcessManager/WebServerReadinessProbeTrait.php]: Because it instead uses a socket, I thought it might be a much better idea

Comment (dunglas) [reply src/ProcessManager/WebServerReadinessProbeTrait.php]: I'm 👎 to use a socket. Using a HTTP client makes sure that the underlying API speaks HTTP. (A server supporting another protocol may already use this port).

**Comment by Pierstoval** [comment]
As #28 is merged, the bugfix part is OK.

For the rest, we discussed with @dunglas on Slack and for now the socket part is not something we really need.

Let's close 😄
